### PR TITLE
refactor logger

### PR DESCRIPTION
**Pull request by VincentBailly**
- remove global states
- make it possible to configure different output than the console
- better type safety for timing
- various responsibilities of the big logger object implemented in small and single purpose modules
- remove some implementation methods that were exposed at the interface of logger

### CODE REVIEW

**Comment by VincentBailly** [comment]
TODO: use the generic logging for output emitted by execa

File: packages/logger/src/logger.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    `perf-${performanceReportData.name}-${performanceReportData.timestamp}.json`
+  );
+}
+
+export type Logger = ConsoleLogger & {
+  setName(name: string): void;
+  setHash(hash: string): void;
+  setCacheProvider(cacheProvider: string): void;
+  setHit(hit: boolean): void;
+  setTime(type: Times): { stop(): void };
```
[the remark this answers is not in the document]
Comment (bweggersen) [reply packages/logger/src/logger.ts:44]: Or did you figure out that every use of the timer was related to `setTime`, so you just made it a part of that?

File: packages/hasher/src/index.ts [hunk 2]
```diff
@@ -13,6 +11,7 @@ import {
   findWorkspacePath,
   WorkspaceInfo
 } from "./yarnWorkspaces";
+import { Logger } from "backfill-logger";
```
[the remark this answers is not in the document]
Comment (VincentBailly) [reply packages/hasher/src/index.ts:14]: I was expecting the git hooks to do this, lets see if we cannot set that up.

File: packages/backfill/src/index.ts [hunk 3]
```diff
@@ -35,7 +36,7 @@ export async function backfill(
   } = config;
 
   logger.setName(name);
-  logger.setMode(mode);
+  logger.setMode(mode, mode === "READ_WRITE" ? "info" : "verbose");
```
[the remark this answers is not in the document]
Comment (VincentBailly) [reply packages/backfill/src/index.ts:39]: this condition was implemented in performance-logger before

File: packages/logger/src/logger.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (VincentBailly) [reply packages/logger/src/logger.ts:44]: All use of the timer are used in the report produced to disk.
It is not clear to me whether you suggest to add a method or to rename setTime()?

File: packages/logger/src/logger.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+  return {
+    pipeProcessOutput(stdout: Readable | null, stderr: Readable | null): void {
+      stdout &&
+        stdout.on("data", chunk =>
+          consoleLogger.consoleOverride.info(chunk.toString())
+        );
+      stderr &&
+        stderr.on("data", chunk =>
+          consoleLogger.consoleOverride.error(chunk.toString())
```
[the remark this answers is not in the document]
Comment (VincentBailly) [reply packages/logger/src/logger.ts:69]: I don't think so, the official documentation does not https://nodejs.org/api/child_process.html#child_process_subprocess_stdout

File: packages/hasher/src/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (VincentBailly) [reply packages/hasher/src/index.ts:14]: I will add es-lint to backfill in the next PR to not have massive merge conflict with this one.

File: packages/logger/src/logger.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (bweggersen) [reply packages/logger/src/logger.ts:43]: Let's just leave it the way you changed it :)

**Comment by bweggersen** [comment]
Perfect! Ship it!

**Pull request merged**